diff --git a/components/lwip/apps/sntp/sntp.c b/components/lwip/apps/sntp/sntp.c
index 6bee032f..48edac95 100644
--- a/components/lwip/apps/sntp/sntp.c
+++ b/components/lwip/apps/sntp/sntp.c
@@ -198,6 +198,19 @@ static ip_addr_t sntp_last_server_address;
 static u32_t sntp_last_timestamp_sent[2];
 #endif /* SNTP_CHECK_RESPONSE >= 2 */
 
+/**
+ * Weak callbacks to implement setting system time
+ */
+extern void __attribute__((weak)) sntp_setsystemtime_us(time_t t, suseconds_t us)
+{
+  SNTP_SET_SYSTEM_TIME_US(t, us);
+}
+
+extern void __attribute__((weak)) sntp_setsystemtime(time_t t)
+{
+  SNTP_SET_SYSTEM_TIME(t);
+}
+
 /**
  * SNTP processing of received timestamp
  */
@@ -214,14 +227,14 @@ sntp_process(u32_t *receive_timestamp)
 
 #if SNTP_CALC_TIME_US
   u32_t us = ntohl(receive_timestamp[1]) / 4295;
-  SNTP_SET_SYSTEM_TIME_US(t, us);
+  sntp_setsystemtime_us(t, us);
   /* display local time from GMT time */
   LWIP_DEBUGF(SNTP_DEBUG_TRACE, ("sntp_process: %s, %"U32_F" us", ctime(&tim), us));
 
 #else /* SNTP_CALC_TIME_US */
 
   /* change system time and/or the update the RTC clock */
-  SNTP_SET_SYSTEM_TIME(t);
+  sntp_setsystemtime(t);
   /* display local time from GMT time */
 
   LWIP_DEBUGF(SNTP_DEBUG_TRACE, ("sntp_process: %s", ctime(&tim)));
